DS-157: Transfer AVAX Tokens Between Chains (DON'T MERGE) (#782)
ava-labs/avalanche-docs: docs/quickstart/cross-chain-transfers.md

# Cross Chain Transfers

## Introduction

> AVAX tokens exist on the X-Chain, where they can be traded, on the P-Chain, where they can be provided as a stake when validating the Primary Network, and on the C-Chain, where they can be used in smart contracts or to pay for gas. In this tutorial, we’ll send AVAX tokens between the X-Chain and C-Chain.

Reviewer: ```suggestion
This article shows how to transfer AVAX tokens programmatically between any two chains of the Primary Network.  
```

## Requirements

> You've completed [Run an Avalanche Node](../nodes/build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../overview/getting-started/avalanche-platform.md). You are familiar with and have cloned the [AvalancheJS](https://github.com/ava-labs/AvalancheJS) repo.

Reviewer: Do they have to clone AvalancheJS? Can they just install and use it?

Author: They would have to clone it if they want to change their Avalanche network config

or 

A user could add the Avalanche pakcage to their project , edit the Avalanche network config in the export script 'node_modules/avalanche/exampes/avm/buildExport...', then run it with ts-node

Reviewer: Please describe these in the article.

> In order to send AVAX, you need to have some AVAX! You can get real AVAX by buying it on an exchange, or you can get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax-test.network), which is a free and easy way to get to play around with Avalanche.

Reviewer: ```suggestion
In order to send AVAX, you need to have some AVAX! You can get real AVAX by buying it on an exchange, or you can get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche.
```

| [**PlatformVM : Import Avax from X-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildImportTx-XChain.ts) | _X Chain >> P Chain_ |
| [**PlatformVM : Export Avax to X-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildExportTx-XChain.ts) | _P Chain >> X Chain_ |
| [**AVM : Import Avax from P-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildImportTx-PChain.ts) | _P Chain >> X Chain_ |
| [**EVM : Export Avax to P-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-pchain.ts) | _C Chain >> P Chain_ |
| [**PlatformVM : Import Avax from C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildImportTx-CChain.ts) | _C Chain >> P Chain_ |
| [**PlatformVm : Export Avax to C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildExportTx-CChain.ts) | _P Chain >> C Chain_ |
| [**EVM : Import Avax from P-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildImportTx-PChain.ts) | _P Chain >> C Chain_ |

> 

Reviewer: Can you group these based on source chain name, and use X/P/C-Chain as name?
* X-Chain
* P-Chain
* C-Chain

> ### Step 2 - Modify your export

Reviewer: ```suggestion
### Step 2 - Modify Your Export
```
Title case for titles.

:::info

The script in this tutorial uses **X Chain <-> C Chain** transfers. Depending on which chains you want to interact with, the scripts you use could be different. Please check the table above for more information.

:::

Select the **examples/avm** folder to view the AvalancheJS X-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [```avm/buildExportTx-cchain-avax.ts```](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts)

![Image for post](/img/ajs-export-cchain-res.png)

#### Modify your Avalanche network configuration

```js

> const ip: string = "localhost" | "custom"

Reviewer: ```suggestion
const ip: string = "localhost" | "api.avax.network" | "api.avax-test.network"
```

Reviewer: Please explain which IP to use for what scenario.

Reviewer: This was not addressed.

Reviewer: This is still not addressed

```
```js
const port: number = 9650 | custom
```
```js
const protocol: string = "http" | "https"
```

```js

> const networkID: number =

Reviewer: Please explain what this number is?

Reviewer: This was not addressed.

| [**PlatformVM : Import Avax from C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildImportTx-CChain.ts) | _C Chain >> P Chain_ |
| [**PlatformVm : Export Avax to C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildExportTx-CChain.ts) | _P Chain >> C Chain_ |
| [**EVM : Import Avax from P-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildImportTx-PChain.ts) | _P Chain >> C Chain_ |

### Step 2 - Modify your export

:::info

The script in this tutorial uses **X Chain <-> C Chain** transfers. Depending on which chains you want to interact with, the scripts you use could be different. Please check the table above for more information.

:::

Select the **examples/avm** folder to view the AvalancheJS X-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [```avm/buildExportTx-cchain-avax.ts```](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts)

![Image for post](/img/ajs-export-cchain-res.png)

#### Modify your Avalanche network configuration

```js
const ip: string = "localhost" | "custom"
```
```js

> const port: number = 9650 | custom

Reviewer: Please explain which port to use for what scenario.

```
```js
const protocol: string = "http" | "https"
```

```js
const networkID: number =
  0: "custom"
  1: "avax"
  2: "cascade"
  3: "denali"
  4: "everest"
  5: "fuji"
  1337: "custom"
  12345: "local"
  ```

Example:
```js
const ip: string = "localhost"

> const port: number = 13445

Reviewer: Should this be 9650?

const protocol: string = "http"
const networkID: number = 1337
```

> #### Edit the amount of AVAX you want to send:

Reviewer: Title case

By default the script sends the wallet's entire AVAX balance:

```js
const balance: BN = new BN(getBalanceResponse.balance)
const amount: BN = balance.sub(fee)
```

> Change the amount by creating a new _BN_ variable: (```value```) and assigning it a string value (```"10000000000000000"```) 

Reviewer: Please explain that the unit is GWEI and conversion from GWEI to AVAX.

```js
  const value: BN = new BN("10000000000000000")
  const amount: BN = value.sub(fee)
  ```

### Step 3 - Run the Export and Import scripts

```zsh
avalanchejs $ ts-node examples/avm/buildExportTx-cchain-avax.ts
Success! TXID: Rgg2412kaczRYC3taasvG6bYoqG7tBQG6WfacNdumKDKsVWpF

> avalanchejs $ ts-node examples/evm/buildImportTx-xchain.ts               

Reviewer: `examples/evm/buildImportTx-xchain.ts` was not mentioned before. Please explain what it does.

Author: https://github.com/ava-labs/avalanche-docs/blob/ebc232b5fa0f117a69328a08c3a7142613b6c13a/docs/quickstart/cross-chain-transfers.md?plain=1#L237

## Requirements

You've completed [Run an Avalanche Node](../nodes/build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../overview/getting-started/avalanche-platform.md). You are familiar with and have cloned the [AvalancheJS](https://github.com/ava-labs/AvalancheJS) repo.

In order to send AVAX, you need to have some AVAX! You can get real AVAX by buying it on an exchange, or you can get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche.

## Transferring AVAX using AvalancheJS

The easiest way to transfer AVAX between chains is to use [AvalancheJS](https://github.com/ava-labs/AvalancheJS) which is a programmatic way to access and move AVAX.

### Step 1 - Open AvalancheJS

![Image for post](/img/examples.png)

Select the **examples** folder to view the source code for AJS examples. To send AVAX between the Chains see the following:

| Source Chain | Title | Path |
| :--- | :--- | :--- |
| ***X-Chain*** | [**AVM : Export Avax to C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts) | _X Chain >> C Chain_ |

> | ***X-Chain*** | [**EVM : Import Avax from X-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildImportTx-xchain.ts) | _X Chain >> C Chain_ |

Reviewer: Suggestion: let's avoid use the term AVM/EVM/PlatformVM, instead use X-Chain/C-Chain/P-Chain here.

## Introduction

This article shows how to transfer AVAX tokens programmatically between any two chains of the Primary Network.

If you are looking for how to transfter AVAX tokens using the web wallet, please check out [this article](https://support.avax.network/en/articles/6169872-how-to-make-a-cross-chain-transfer-in-the-avalanche-wallet).

## Requirements

You've completed [Run an Avalanche Node](../nodes/build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../overview/getting-started/avalanche-platform.md). You are familiar with the [AvalancheJS](https://github.com/ava-labs/AvalancheJS) repo.

To use AvalancheJS, you can clone the repo or add it to an existing project by running the following command:

```zsh
$ yarn add --dev avalanche
```
For this tutorial we will use a cli to run the example scripts directly from an AvalancheJS directory.

In order to send AVAX, you need to have some AVAX! You can get real AVAX by buying it on an exchange, or you can get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche.

## Transferring AVAX using AvalancheJS

> The easiest way to transfer AVAX between chains is to use [AvalancheJS](https://github.com/ava-labs/AvalancheJS) which is a programmatic way to access and move AVAX.

Reviewer 2: Add that we think using the AvalancheGo's keystore is an anti-pattern because it requires you to keep your keys on a full node which makes them a target for malicious hackers. Instead the recommended way is to use AvalancheJS which allows you to create and sign transactions locally.

Reviewer: Following up this comments, please show in the code where to put the private key.

Author: https://github.com/ava-labs/avalanche-docs/blob/ebc232b5fa0f117a69328a08c3a7142613b6c13a/docs/quickstart/cross-chain-transfers.md?plain=1#L51

| ***C-Chain*** | [**P-Chain : Import Avax from C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildImportTx-CChain.ts) | _C Chain >> P Chain_ |

### Step 2 - Modify Your Export

:::info

The script in this tutorial uses **X Chain <-> C Chain** transfers. Depending on which chains you want to interact with, the scripts you use could be different. Please check the table above for more information.

:::

Select the **examples/avm** folder to view the AvalancheJS X-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [```avm/buildExportTx-cchain-avax.ts```](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts)

![Image for post](/img/ajs-export-cchain-res.png)

#### Modify your Avalanche network configuration

If you are making calls to your Node from another machine, or started your node using [Advanced Configurations](https://docs.avax.network/nodes/build/set-up-node-with-installer#advanced-node-configuration) you might use a custom IP
```js

> const ip: string = "localhost" | "custom" 

Reviewer: ```suggestion
const ip: string = "localhost" | "api.avax.network" | "api.avax-test.network"
```
Please explain which IP to use in what scenario.

> const ip: string = "localhost" | "custom" 

Reviewer 3: If the custom IP is an actual IP, then it might be better to put something like
`const ip: string = "localhost" | "<YOUR-NODE-IP>"`
Otherwise people will put the actual "custom" string.

Also, be careful with things like `"localhost" | "custom"`- this is a valid JS code. It does not convey that people just copy/pasting your instruction should use one or the other.

Don't be afraid to waste screen space, scrolling is cheap. You might present it like:

...you might use a local node:
```js
const ip: string = "localhost"
```
Or a custom node IP:
```js
const ip: string = "<YOUR-NODE-IP_HERE>"
```

```

```js
const port: number = 9650
```
```js

> const protocol: string = "http" | "https"

Reviewer 3: Same here. I bet a nontrivial number of people trying to reproduce this tutorial would just leave that line verbatim, not select http or https and delete the other.

```

> Depending on the networkID which is passed in when instantiating Avalanche the encoded addresses will have a distinctive HRP per each network.

Reviewer: what does HRP stand for?

Author: https://github.com/ava-labs/avalanche-docs/blob/ebc232b5fa0f117a69328a08c3a7142613b6c13a/docs/quickstart/cross-chain-transfers.md?plain=1#L171

```js
const networkID: number =
  0: "custom"
  1: "avax"
  2: "cascade"
  3: "denali"
  4: "everest"
  5: "fuji"
  1337: "custom"
  12345: "local"
  ```

Example:
```js
const ip: string = "localhost"
const port: number = 9650
const protocol: string = "http"
const networkID: number = 5

> ```

Reviewer 3: This is the best way, provide concrete examples. But the above point stands, avoid using programming language constructs to convey instructions, people can easily misunderstand your meaning.

#### Edit The Amount Of AVAX You Want To Send:
By default the script sends the wallet's entire AVAX balance:

```js
const balance: BN = new BN(getBalanceResponse.balance)
const amount: BN = balance.sub(fee)
```

> Change the amount by creating a new _BN_ variable: ```value``` and assigning it a string value ```"10000000000000000"``` (.01 AVAX) 

Reviewer: ```suggestion
Change the amount by creating a new _BN_ variable: ```value``` and assigning it a string value ```"10000000000000000"``` (.01 AVAX) as an example. 
```

```js
  const value: BN = new BN("10000000000000000")
  const amount: BN = value.sub(fee)
  ```

We must pass the value in WEI format. For reference, 10 ** 18 WEI = 1 AVAX

You can use the [snowtrace unit converter](https://snowtrace.io/unitconverter) to view more unit conversions

> 

Reviewer 2: Add a section that points out which address as the actual AVAX balance and is paying the fee. Deep link to [this line](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) in the export script and point out that `DefaultLocalGenesisPrivateKey` is [an AvalancheJS constant](https://github.com/ava-labs/avalanchejs/blob/master/src/utils/constants.ts#L134-L135) and this is the bech32 address which is derived from that private key: `X-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p`

That key is pre-funded on the local network when using Avalanche Network Runner. It has the following AVAX in the genesis block/vertex of each chain on the primary subnet

* X-Chain: 300000000000000000 nAVAX
* P-Chain: 30000000000000000 nAVAX
* C-Chain: 50000000000000000000000000 nAVAX

Author: Added Earlier

https://github.com/ava-labs/avalanche-docs/blob/ebc232b5fa0f117a69328a08c3a7142613b6c13a/docs/quickstart/cross-chain-transfers.md?plain=1#L56

  },
  credentials: [
    {
      fxID: 'spdxUxVJQbX85MGxMHbKw1sHxMnSqJ3QBzDyDYEP3h6TLuxqQ',
      credential: [Object]
    }
  ]
}
```

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction:

- Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts)

- Make necessary changes as above for private key and network settings.

- You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (0.01 AVAX)

The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.

```ts
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));
fee = fee.add(new BN(1e6));
```

:::tip

> When exporting AVAX, be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (0.001 AVAX) may cause the import txn to fail.

Reviewer: How does this `Sending less than `1e6` .......` work with line 229 
`let fee: BN = baseFee.div(new BN(1e9));`
which is less than 1e6 ? @Julian-dev28

Author: I see 

the fee is less than 1e6 initially but we add the minimum import fee(.001 Avax) on the next line

`fee = fee.add(new BN(1e6));`

will update

:::

Run the export script:

```sh
$ ts-node examples/avm/buildExportTx-cchain-avax.ts
```

This returns:

```sh
Success! TXID: Rgg2412kaczRYC3taasvG6bYoqG7tBQG6WfacNdumKDKsVWpF
```

#### Import the Avax Token From X-Chain to C-Chain

Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To import AVAX to the C-Chain from the X-Chain, select [`evm/buildImportTx-xchain.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildImportTx-xchain.ts)

Copy the [network setting from above](#network-setting) into `evm/buildImportTx-xchain.ts`.

Navigate to this part of the code and ensure that the `cHexAddress`(_Your C-Chain wallet address_) and `private key` are correct:

```ts
const cHexAddress: string = "<YOUR-CCHAIN-WALLET-ADDRESS-HERE>";
const privKey: string = "<YOUR-PRIVATE-KEY-HERE>";
```

Run the import script:

```sh
$ ts-node examples/evm/buildImportTx-xchain.ts
```

This returns:

```sh
Success! TXID: r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1
```

That's it! You've transferred AVAX from the X-Chain to C-Chain!

#### Verify the Transaction

> You can now pass this txID `r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1` into [examples/avm/getTx.ts](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/getTx.ts), plus other similar network settings, then you can run

Reviewer: Which tx ID should be put in `getTx.ts`? will both `r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1` and `Rgg2412kaczRYC3taasvG6bYoqG7tBQG6WfacNdumKDKsVWpF` work?

```zsh
$ ts-node examples/avm/getTx.ts
```

which returns:

```js
{
  unsignedTx: {

>     networkID: 1,

Reviewer: is this output correct? it has `networkID: 1,` where for Fuji the network ID is 5.

Author: fixed

added new example

    blockchainID: '2oYMBNV4eNHyqk2fjjV5nVQLDbtmNJzq5s3qs3Lo6ftnC6FByM',
    outputs: [],
    inputs: [ [Object] ],
    memo: '0x',
    destinationChain: '11111111111111111111111111111111LpoYY',
    exportedOutputs: [ [Object] ]
  },
  credentials: [
    {
      fxID: 'spdxUxVJQbX85MGxMHbKw1sHxMnSqJ3QBzDyDYEP3h6TLuxqQ',
      credential: [Object]
    }
  ]
}
```

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction:

- Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To export AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts)

- Make necessary changes as above for private key and network settings.

- You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (0.01 AVAX)

> The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.

Reviewer: Please describe the code for import in a separate paragraph with `examples/avm/buildImportTx-cchain.ts` so it is much clear instead of lumping them together here.

Reviewer: Is this sentence `The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.` still apply here?

Author: Yes it does

It is informational and describes the fee mechanism

Reviewer: But this section focuses on `Export the Avax Token From C-Chain to X-Chain`, why is Import fee being brought up? This should be moved to the next section `Import the Avax Token From C-Chain to X-Chain`.

Author: We need to pay the import fees while exporting so it does make sense

also this code:

```js
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));
fee = fee.add(new BN(1e6));
```

exists on the export script we are using

Author: Makes more sense in the export from X-Chain section

Moving now

```ts
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));
fee = fee.add(new BN(1e6));
```

:::tip
When exporting AVAX, be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (0.001 AVAX) may cause the import txn to fail.
:::

The following sequence demonstrates how to send AVAX back to the X-Chain:

```zsh
avalanchejs $ ts-node examples/evm/buildExportTx-xchain-avax.ts
Success! TXID: UAez3DTv26qmhKKFDvmQTayaXTPAVahHenDKe6xnUMhJbKuxc

> avalanchejs $ ts-node examples/avm/buildImportTx-cchain.ts

Reviewer: Please separate this file out in its own paragraph and describe in details what needs to be changed before running it.

const privKey: string = "<YOUR-PRIVATE-KEY-HERE>";
```

Run the import script:

```sh
$ ts-node examples/evm/buildImportTx-xchain.ts
```

This returns:

```sh
Success! TXID: r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1
```

That's it! You've transferred AVAX from the X-Chain to C-Chain!

#### Verify the Transaction

You can now pass this txID `r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1` into [examples/avm/getTx.ts](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/getTx.ts), plus other similar network settings, then you can run

```zsh
$ ts-node examples/avm/getTx.ts
```

which returns:

```js
{
  unsignedTx: {
    networkID: 1,
    blockchainID: '2oYMBNV4eNHyqk2fjjV5nVQLDbtmNJzq5s3qs3Lo6ftnC6FByM',
    outputs: [],
    inputs: [ [Object] ],
    memo: '0x',
    destinationChain: '11111111111111111111111111111111LpoYY',
    exportedOutputs: [ [Object] ]
  },
  credentials: [
    {
      fxID: 'spdxUxVJQbX85MGxMHbKw1sHxMnSqJ3QBzDyDYEP3h6TLuxqQ',
      credential: [Object]
    }
  ]
}
```

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction.

> #### Export the Avax Token From C-Chain to X-Chain

Reviewer: I now separated this section into two sections of `Export` and `Import`. Please update the doc following this structure instead of lumping everything in one section.

Author: roger

Make necessary changes as above for private key and network settings.

You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (0.01 AVAX)

The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.

```ts
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));
fee = fee.add(new BN(1e6));
```

:::tip
When exporting AVAX, be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (0.001 AVAX) may cause the import txn to fail.
:::

Run the export script:

```zsh
avalanchejs $ ts-node examples/evm/buildExportTx-xchain-avax.ts
Success! TXID: UAez3DTv26qmhKKFDvmQTayaXTPAVahHenDKe6xnUMhJbKuxc
```

> #### Import the Avax Token From C-Chain to X-Chain

Reviewer: Please describe in details what needs to be changed to `examples/avm/buildImportTx-cchain.ts` in this section.

Author: Roger

Reviewer: which file?

    exportedOutputs: [ [Object] ]
  },
  credentials: [
    {
      fxID: 'spdxUxVJQbX85MGxMHbKw1sHxMnSqJ3QBzDyDYEP3h6TLuxqQ',
      credential: [Object]
    }
  ]
}
```

#### Import the Avax Token From X-Chain to C-Chain

Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To import AVAX to the C-Chain from the X-Chain, select [`evm/buildImportTx-xchain.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildImportTx-xchain.ts)

Copy the [network setting from above](#network-setting) into `evm/buildImportTx-xchain.ts`.

Navigate to this part of the code and ensure that the `cHexAddress`(_Your C-Chain wallet address_) and `private key` are correct:

```ts
const cHexAddress: string = "<YOUR-CCHAIN-WALLET-ADDRESS-HERE>";
const privKey: string = "<YOUR-PRIVATE-KEY-HERE>";
```

Run the import script:

```sh
$ ts-node examples/evm/buildImportTx-xchain.ts
```

This returns:

```sh
Success! TXID: ft33QHCH542wWUwrjdVf9APtAjPWqzqbqE43nCRECdbzhzVub
```

That's it! You've transferred AVAX from the X-Chain to C-Chain!

> You can verify this TX by copy / pasting the import TXID into [Avascan](https://testnet.avascan.info/blockchain/c/tx/ft33QHCH542wWUwrjdVf9APtAjPWqzqbqE43nCRECdbzhzVub)

Reviewer: ```suggestion
You can verify this TX by copy / pasting the import TXID into [Avascan](https://testnet.avascan.info/blockchain/c/tx/ft33QHCH542wWUwrjdVf9APtAjPWqzqbqE43nCRECdbzhzVub).
```

> You can verify this TX by copy / pasting the import TXID into [Avascan](https://testnet.avascan.info/blockchain/c/tx/ft33QHCH542wWUwrjdVf9APtAjPWqzqbqE43nCRECdbzhzVub)

Reviewer: This example shows of importing 9.999 [AVAX](https://testnet.avascan.info/blockchain/x/asset/U8iRqJoiJm8xZHAacmvYyZVwqQx6uDNtQeP3CQ6fcgQk3JqnK) while this examples uses value of 0.01 avax for transfer. Please make them consistent.

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction.

#### Export the Avax Token From C-Chain to X-Chain

> Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To export AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts)

Reviewer: ```suggestion
Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To export AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts).
```

Make necessary changes as above for private key and network settings.

You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (0.01 AVAX)

The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.

```ts
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));
fee = fee.add(new BN(1e6));
```

:::tip
You do not need to adjust `fee`, you will however adjust `avaxAmount`, the variable that the sample code uses to define the amount of AVAX that the user will send. Be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (0.001 AVAX) may cause the import txn to fail.
:::

Run the export script:

```zsh
avalanchejs $ ts-node examples/evm/buildExportTx-xchain-avax.ts
Success! TXID: UAez3DTv26qmhKKFDvmQTayaXTPAVahHenDKe6xnUMhJbKuxc
```

#### Import the Avax Token From C-Chain to X-Chain
Before we run the example import script, we need to make some changes to the code:

> 1. Change the [Network Setting](./cross-chain-transfers.md#network-setting) to meet [Local](./cross-chain-transfers.md#local-workflow), [Fuji](./cross-chain-transfers.md#network-setting), or [Mainnet](./cross-chain-transfers.md#mainnet-workflow) network requirements.

Reviewer: why do you mention Mainnet here?

Reviewer: Also `#network-setting` is enough for the linking.

Reviewer: Please this whole section is under `Fuji Workflow`. Please remove all references to Local and Mainnet here.

> 2. Import your Private Key by following the steps listed [here](./cross-chain-transfers.md#private-key).

Reviewer: Is the fee part needs to be adjusted? If so, how?

> 2. Import your Private Key by following the steps listed [here](./cross-chain-transfers.md#private-key).

Reviewer: ```suggestion
2. Import your Private Key by following the steps listed [here](#private-key).
```

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction.

#### Export the Avax Token From C-Chain to X-Chain

Select the [**`examples/evm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/evm) folder to view the AvalancheJS C-Chain examples. To export AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts).

Make necessary changes as above for private key and network settings.

You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (0.01 AVAX)

The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs stay after getting exported but before being imported.

```ts
let avaxAmount: BN = new BN(1e7);
let fee: BN = baseFee.div(new BN(1e9));

> fee = fee.add(new BN(1e6));

Reviewer: These code does not match anymore https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts#L46

Author: Right

We are blocked by the merge to master

the changes are in development https://github.com/ava-labs/avalanchejs/blob/c7d47ef89dc32b8415d9d41dd686f4038199921a/examples/evm/buildExportTx-xchain-avax.ts#L46

Reviewer: which PR is this?

Author: Yes, this is the one that fixes the issue

Reviewer: Does this mean that this doc PR needs to wait until after next AvalancheJS release?

```

:::tip

> You do not need to adjust `fee`, you will however adjust `avaxAmount`, the variable that the sample code uses to define the amount of AVAX that the user will send. Be sure to send enough to support import fees (constant .001  AVAX or 1e6 gwei). Sending less than 0.001 AVAX may cause the import txn to fail.

Reviewer: ```suggestion
Fees on X-Chain and P-Chain are fixed for Export/Import of AVAX as constant 0.001 AVAX or 1e6 gwei (see [here](./transaction-fees.md#fee-schedule)). 
```

> You do not need to adjust `fee`, you will however adjust `avaxAmount`, the variable that the sample code uses to define the amount of AVAX that the user will send. Be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (0.001 AVAX) may cause the import txn to fail.

Reviewer: ```suggestion
You do not need to adjust `fee`, you will however adjust `avaxAmount`, the variable that the sample code uses to define the amount of AVAX that the user will send. Be sure to send enough to support import fees (constant .001  AVAX or 1e6 gwei). Sending less than 0.001 AVAX may cause the import txn to fail.
```

# Transfer AVAX Tokens Between Chains

## Introduction

This article shows how to transfer AVAX tokens programmatically between any two chains of the Primary Network.

If you are looking for how to transfter AVAX tokens using the web wallet, please check out [this article](https://support.avax.network/en/articles/6169872-how-to-make-a-cross-chain-transfer-in-the-avalanche-wallet).

## Requirements

You've completed [Run an Avalanche Node](../nodes/build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../overview/getting-started/avalanche-platform.md). You are familiar with the [AvalancheJS](https://github.com/ava-labs/AvalancheJS) repo.

To use AvalancheJS, you can clone the repo:
```zsh
$ git clone https://github.com/ava-labs/avalanchejs.git
```

or add it to an existing project:

```zsh
$ yarn add --dev avalanche
```
For this tutorial we will use a cli to run the example scripts directly from an AvalancheJS directory.

> In order to send AVAX, you need to have some AVAX! You can get real AVAX by buying it on an exchange, or you can get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche.

Reviewer: ```suggestion
In order to send AVAX, you need to have some AVAX!  You can use a pre-funded account on local network or get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche. After getting comfortable with your code, you can run the code on Mainnet by making necessary changes. 
```

## Transferring AVAX using AvalancheJS

The easiest way to transfer AVAX between chains is to use [AvalancheJS](https://github.com/ava-labs/AvalancheJS) which is a programmatic way to access and move AVAX.

AvalancheJS allows you to create and sign transactions locally which is why it is the recommended way to transfer AVAX between chains. We are moving away from using AvalancheGo's keystore because it requires you to keep your keys on a full node which makes them a target for malicious hackers.

You can manage private keys you use with AvalancheJS directly in the AvalancheJS directory by doing the following:

Rename [``secrets.example``](https://github.com/ava-labs/avalanchejs/blob/master/examples/secrets.example) to ``secrets.json``

Add your Private Key as an object
```json
{
  "newPassword": "newPassword",
  "oldPassword": "oldPassword",
  "password": "password",
  "token": "token",
  "user": "user",

>   "privateKey": "PrivateKey-ewoqjP7PxY4yr3iLTpLisriqt94hdyDFNgchSxGGztUrTXtNN"

Reviewer: are newPassword/oldPassword/password/token/user needed for this tutorial? If not, please remove them.

Reviewer 3: I believe those were needed for the Keystore API, which we don't use any more. So definitely no need to have them.

}
```

Import your private key into your project
```js
import { privateKey } from "../secrets.json"
```

> By default, the scripts use the AvalancheJS constant, [```DefaultLocalGenesisPrivateKey```](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) as ```privKey``` to  to sign transactions. 

Reviewer: ```suggestion
By default, the scripts use the AvalancheJS constant, [```DefaultLocalGenesisPrivateKey```](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) as ```privKey``` to sign transactions. 
```

```js
const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`
```

The bech32 address which derived from this private key is : ```X-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p```

That key is pre-funded on the local network when using Avalanche Network Runner and has the following AVAX in the genesis block/vertex of each chain on the primary subnet:

- X-Chain: 300000000000000000 nAVAX

- P-Chain: 30000000000000000 nAVAX

- C-Chain: 50000000000000000000000000 nAVAX

You can simply replace this with the imported private key:
```js

> const privKey: string = privateKey

Author: Are you referring to the Image?

> const privKey: string = privateKey

Reviewer: No, I am referring the line number 53/59/76

> const privKey: string = privateKey

Author: Are you referring to the image?

> const privKey: string = privateKey

Reviewer: wasn't this question asked one hour ago and I have answered above?
<img width="1164" alt="Screen Shot 2022-05-31 at 4 20 40 PM" src="https://user-images.githubusercontent.com/12171681/171293706-05223ef8-28a0-4783-9866-4b6a3e56eea7.png">

> const privKey: string = privateKey

Author: I think this is cleared up with [this section](https://github.com/ava-labs/avalanche-docs/blob/5a6f8558cde3927f5b8881b88c0c6f6368e6403a/docs/quickstart/cross-chain-transfers.md#key-management)

```

Or to connect directly to the [Avalanche public server for Fuji Testnet](../apis/avalanchego/public-api-server.md) server:

```js
const port: number = 443;
```

**Protocol**

To use a local node started with [`--network-id=fuji`](../nodes/maintain/avalanchego-config-flags.md#network-id):

```js
const protocol: string = "http";
```

Or to connect directly to the [Avalanche public server for Fuji Testnet](../apis/avalanchego/public-api-server.md) server:

```js
const protocol: string = "https";
```

**Network ID**

Depending on the networkID passed in when instantiating Avalanche, the encoded addresses used will have a distinctive Human Readable Part(HRP) per each network.

_Example Address: 5 - X-`fuji`19rknw8l0grnfunjrzwxlxync6zrlu33yxqzg0h_

For Fuji Testnet, 5 is the correct value to use.

```js
const networkID: number = 5;
```

To learn more  about  encoded addresses, click [here](../apis/avalanchejs/manage-x-chain-keys.md#encode-bech32-addresses).

**Set the Correct Amount to Send:**

By default the scripts send the wallet's entire AVAX balance:

```js
const balance: BN = new BN(getBalanceResponse.balance);
const amount: BN = balance.sub(fee);
```

> To send a different amount, please replace the code above with the following. Below sets a new value of 0.01 AVAX (10000000000000000 GWEI). Value is set in WEI format where 10 \*\* 18 WEI = 1 AVAX

Reviewer: is 10000000000000000GWEI 0.01AVAX or `10000000` (.01 AVAX) (Line 250)?

Author: 10000000 GWEI confirmed

Thought it was WEI at first

```js
const value: BN = new BN("10000000000000000");
const amount: BN = value.sub(fee);
```

:::tip
Snowtrace provides a [unit converter](https://snowtrace.io/unitconverter) between different units
:::

#### Execute The Export And Import Scripts

To transfer the specified amount token from X-Chain to C-Chain, the token needs to be first exported from the X-Chain, then imported to C-Chain. 

Run the export script:
```sh
avalanchejs $ ts-node examples/avm/buildExportTx-cchain-avax.ts
```

This returns:
```sh
Success! TXID: Rgg2412kaczRYC3taasvG6bYoqG7tBQG6WfacNdumKDKsVWpF
```

Run the import script:
```sh

> avalanchejs $ ts-node examples/evm/buildImportTx-xchain.ts

Reviewer: Is there any changes needed in buildImportTx-xchain.ts before running it? Please list out all the changes and what should be done here.

      credential: [Object]
    }
  ]
}
```

### Transfer from the C-Chain to the X-Chain

To return the AVAX back to the X-Chain, you need to do the transfer in the opposite direction:

* Select the **`examples/evm`** folder to view the AvalancheJS C-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [`evm/buildExportTx-xchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts)

* You can change the amount of AVAX to send by editing the _BN_ variable: `avaxAmount`. The sample code assigns this as `1e7` or `10000000` (.01 AVAX)

The fee here will only be for exporting the asset. The import fees will be deducted from the UTXOs present on the Exported Atomic Memory, a memory location where UTXOs lie after getting exported but before being imported. If there is only a single UTXO, then it will be deducted from it.

```ts
let avaxAmount: BN = new BN(1e7)
let fee: BN = baseFee.div(new BN(1e9))
fee = fee.add(new BN(1e6))
```

:::tip
When exporting AVAX, be sure to send enough to support import fees (constant .001 AVAX). Sending less than `1e6` or `1000000` (.001 AVAX) may cause the import txn to fail._
:::

> Swap source and destination chains by running the [C-Chain Export](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-xchain-avax.ts) and [X-Chain Import](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildImportTx-cchain.ts) scripts.

Reviewer: What does `Swap source and destination chains` mean here?

Author: Updated with a new line 

Please let me know it makes sense

```zsh
avalanchejs $ ts-node examples/evm/buildExportTx-xchain-avax.ts
Success! TXID: UAez3DTv26qmhKKFDvmQTayaXTPAVahHenDKe6xnUMhJbKuxc

> avalanchejs $ ts-node examples/avm/buildImportTx-cchain.ts

Reviewer: Any changes needed in examples/avm/buildImportTx-cchain.ts before running it?

Author: Should be good now

https://github.com/ava-labs/avalanche-docs/blob/328a380d08ef77bf20d3898bcccd8b7ecd991aa8/docs/quickstart/cross-chain-transfers.md?plain=1#L295

You can manage the private keys you use with AvalancheJS directly in the AvalancheJS directory by doing the following:

1. Rename [`secrets.example`](https://github.com/ava-labs/avalanchejs/blob/master/examples/secrets.example) to `secrets.json`

2. Add your Private Key as an object.

```json
// secrets.json
{
  "privateKey": "<YOUR-PRIVATE-KEY-HERE>"
}
```

3. Import your private key into the example script.

```js
// buildExportTx-cchain-avax.ts
import { privateKey } from "../secrets.json";
```

By default, the scripts use the AvalancheJS constant, [`DefaultLocalGenesisPrivateKey`](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) as `privKey` to sign transactions.

```js
// buildExportTx-cchain-avax.ts
const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
```

To run AvalancheJS scripts on Mainnet or Fuji, you can simply replace this with the imported private key:

```js
// buildExportTx-cchain-avax.ts
const privKey: string = privateKey;
```

### Step 1 - Open AvalancheJS

Open your AvalancheJS directory and select the **`examples`** folder to view the source code for the examples scripts.

> ![Image for post](/img/examples.png)

Reviewer: No need for a screenshot here.

### Step 2 - Modify Your Export

:::info

The script in this tutorial uses **X Chain <-> C Chain** transfers. Depending on which chains you want to interact with, the scripts you use could be different. Please check the table above for more information.

:::

<br></br>

Select the **`examples/avm`** folder to view the AvalancheJS X-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [`avm/buildExportTx-cchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts)

![Image for post](/img/ajs-export-cchain-res.png)

**Modify Your Avalanche Network Configuration**

Depending on how you started your node, you might use custom settings for your Avalanche Network

_Note: See [Advanced Configurations](https://docs.avax.network/nodes/build/set-up-node-with-installer#advanced-node-configuration) to learn more about custom Node configurations_

Below are the following objects in your Avalanche Network:

```js

> const ip: string = "172.11.111.123";

Reviewer: what is 172.11.111.123?

const port: number = 9650;
const protocol: string = "http";
const networkID: number = 5;
```

**IP Address**

> If you're using a local node:

Reviewer: This is under the title of "Typical Fuji Workflow", please just discuss Fuji related values.

```js
const ip: string = "localhost";
```

Or a custom IP address:

```js
const ip: string = "<YOUR-NODE-IP-ADDRESS-HERE>";
```

You can also perform this tutorial on the Avalanche Mainnet or Fuji Testnet by connecting to the [Avalanche Public API](../apis/avalanchego/public-api-server.md) server:

Mainnet:

```js
const ip: string = "api.avax.network";
```

Fuji Testnet:

```js
const ip: string = "api.avax-test.network";
```

**Port Number**

Default port number:

```js

> const port: number = 9650;

Reviewer: Is 9650 still the correct value here?

Reviewer: unless you are running a local node. If you are using the public api server for testnet, it should be 443.

```

**Protocol**

> Local node:

Reviewer: only Fuji related value, please.

```js
const protocol: string = "http";
```

[Avalanche Public API](../apis/avalanchego/public-api-server.md):

```js
const protocol: string = "https";
```

**Network ID**

Depending on the networkID which is passed in when instantiating Avalanche, the encoded addresses used will have a distinctive Human Readable Part(HRP) per each network.

_Example Address: 5 - X-`fuji`19rknw8l0grnfunjrzwxlxync6zrlu33yxqzg0h_

Mainnet:

```js
const networkID: number = 1;
```

Fuji Testnet:

```js
const networkID: number = 5;
```

You can learn more about our encoded addresses [here](cross-chain-transfers.md#network-id)

> **Edit the amount of AVAX you want to send:**

Reviewer: I felt is it very redundant to put every thing again in Fuji flow. Please rewrite it and I will review again.

Author: https://github.com/ava-labs/avalanche-docs/pull/782#issuecomment-1142300973

Ok we are moving away from this?

Reviewer: Adding my reply from Slack to keep the discussion whole here:

No, I would like the doc to follow that to be clear on each flow, by separating out local and fuji.

Current doc has repeated almost the same in both Local and Fuji which is very redundant and unclear

This is the main point: `As a developer, I would like to follow through this tutorial from start to end to make it work on one environment first, before I try on another.`

## Getting Started

To use AvalancheJS, you can clone the repo:

```zsh
$ git clone https://github.com/ava-labs/avalanchejs.git
```

or add it to an existing project:

```zsh
$ yarn add --dev avalanche
```

For this tutorial we will use a CLI to run the example scripts directly from an AvalancheJS directory.

In order to send AVAX, you need to have some AVAX! You can use a pre-funded account on local network or get testnet AVAX from the [AVAX Test Faucet](https://faucet.avax.network), which is a free and easy way to get to play around with Avalanche. After getting comfortable with your code, you can run the code on Mainnet by making necessary changes.

## Transferring AVAX using AvalancheJS

The easiest way to transfer AVAX between chains is to use [AvalancheJS](https://github.com/ava-labs/AvalancheJS) which is a programmatic way to access and move AVAX.

AvalancheJS allows you to create and sign transactions locally which is why it is the recommended way to transfer AVAX between chains. We are moving away from using AvalancheGo's keystore because it requires you to keep your keys on a full node which makes them a target for malicious hackers.

> You will find both a [Fuji Workflow](cross-chain-transfers.md#typical-fuji-workflow) and [Avalanche Network Runner Workflow](cross-chain-transfers.md#typical-avalanche-network-runner-workflow) in this tutorial.

Reviewer: Both links are broken here.

## Typical Avalanche Network Runner Workflow

First confirm you have the latest AvalancheGo built.

```text
$ cd /path/to/avalanchego
$ git fetch -p
$ git checkout master
$ ./scripts/build.sh
```

(Note that you can also [download pre-compiled AvalancheGo binaries](https://github.com/ava-labs/avalanchego/releases) rather than building from source.)

Confirm you have Avalanche Network Runner installed by following the steps listed [here](create-a-local-test-network.md#installation)

Start Avalanche Network Runner and run a script to start a new local network.

> ### Step 1 - Start the server

Reviewer: Title case for this and all titles.

```text
$ cd /path/to/Avalanche-Network-Runner
$ avalanche-network-runner server \
--log-level debug \
--port=":8080" \
--grpc-gateway-port=":8081"
```

### Step 2 - Start a New Avalanche Network with Five Nodes (a cluster)

```bash
# replace execPath with the path to AvalancheGo on your machine
# e.g., ${HOME}/go/src/github.com/ava-labs/avalanchego/build/avalanchego
$ AVALANCHEGO_EXEC_PATH="avalanchego"
```

```bash
$ avalanche-network-runner control start \
--log-level debug \
--endpoint="0.0.0.0:8080" \
--number-of-nodes=5 \
--avalanchego-path ${AVALANCHEGO_EXEC_PATH}
```

Now you're running a local Avalanche network with 5 nodes.

### Step 3 - Open AvalancheJS

> Open your AvalancheJS directory and select the **`examples`** folder to view the source code for the examples scripts.

Reviewer: Please link the `example` to the github location `https://github.com/ava-labs/avalanchejs/tree/master/examples`

| **_C-Chain_** | [**X-Chain : Import Avax from C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildImportTx-cchain.ts)        | _C Chain >> X Chain_ |
| **_C-Chain_** | [**C-Chain : Export Avax to P-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/evm/buildExportTx-pchain.ts)          | _C Chain >> P Chain_ |
| **_C-Chain_** | [**P-Chain : Import Avax from C-Chain**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildImportTx-CChain.ts) | _C Chain >> P Chain_ |

### Step 4 - Modify Your Export

:::info

The script in this tutorial uses **X Chain <-> C Chain** transfers. Depending on which chains you want to interact with, the scripts you use could be different. Please check the table above for more information.

:::

<br></br>

Select the **`examples/avm`** folder to view the AvalancheJS X-Chain examples. To send AVAX from the X-Chain to the C-Chain, select [`avm/buildExportTx-cchain-avax.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts)

![Image for post](/img/ajs-export-cchain-res.png)

> ### Modify your Avalanche network configuration

Reviewer: This looks weird
<img width="326" alt="Screen Shot 2022-06-02 at 7 50 20 AM" src="https://user-images.githubusercontent.com/12171681/171644731-896df947-067e-4bee-86ab-884f48f5ebdc.png">

Depending on how you started your node, you might use custom settings for your Avalanche Network

You might notice that Avalanche Network Runner produces new ports for each node during node cluster deployment. Be sure to change configurations accordingly with each new cluster.

_Note: See [Advanced Configurations](https://docs.avax.network/nodes/build/set-up-node-with-installer#advanced-node-configuration) to learn more about custom Node configurations_

> Below are the following objects in your Avalanche Network:

Reviewer: ```suggestion
Below are the settings for running code with local network:
```

```js
const ip: string = "localhost";
const port: number = 9650;
const protocol: string = "http";
const networkID: number = 1337;
```

**IP Address**

> You Node's IP address

Reviewer: ```suggestion
```

> By default, Avalanche Network Runner uses localhost.

Reviewer: ```suggestion
By default, AvalancheJS  uses localhost.
```

```js
const ip: string = "localhost";
```

**Port Number**

As stated above, Avalanche Network Runner produces new port numbers for each node cluster.

```js
const port: number = "<YOUR-NODE-PORT-HERE>";
```

**Protocol**

The protocol we use to connect to the node's IP

```js
const protocol: string = "http";
```

#### **Network ID**

Depending on the networkID which is passed in when instantiating Avalanche, the encoded addresses used will have a distinctive Human Readable Part(HRP) per each network.

_Example Address: 1337 - X-`custom`19rknw8l0grnfunjrzwxlxync6zrlu33yeg5dya_

Mainnet:

```js
const networkID: number = 1;
```

Fuji Testnet:

```js
const networkID: number = 5;
```

> Custom Network:

Reviewer: is this Custom Network or Local Network?

```js
const networkID: number = 1337;
```

We use Bech32 to encode our addresses. Each Bech32 address is composed of the following components:

- A Human-Readable Part (HRP).
- The number “1” is a separator (the last digit 1 seen is considered the separator).
- Base-32 encoded string for the data part of the address (the 20-byte address itself).
- A 6-character base-32 encoded error correction code using the BCH algorithm.

AvalancheJS has address encoding for the following networks

```text
0 - X-custom19rknw8l0grnfunjrzwxlxync6zrlu33yeg5dya
1 - X-avax19rknw8l0grnfunjrzwxlxync6zrlu33y2jxhrg
2 - X-cascade19rknw8l0grnfunjrzwxlxync6zrlu33ypmtvnh
3 - X-denali19rknw8l0grnfunjrzwxlxync6zrlu33yhc357h
4 - X-everest19rknw8l0grnfunjrzwxlxync6zrlu33yn44wty
5 - X-fuji19rknw8l0grnfunjrzwxlxync6zrlu33yxqzg0h
1337 - X-custom19rknw8l0grnfunjrzwxlxync6zrlu33yeg5dya
12345 - X-local19rknw8l0grnfunjrzwxlxync6zrlu33ynpm3qq

> ```

Reviewer: Is this section needed? I feel this just introduces more complexity. Maybe move this to an appendix section.

### Edit The Amount Of AVAX You Want To Send:

By default the scripts send the wallet's entire AVAX balance:

> When using Avalanche Network Runner, we import our [Pre-Generated Private Key](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) to our [`xKeychain`](https://github.com/ava-labs/avalanchejs/blob/46ce89f395133702320a77cba4bb9cb818b48fe8/examples/avm/buildExportTx-cchain-avax.ts#L31) to obtain signers. This means `X-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p` is the `balance` holder and will pay the `fee` to export AVAX.

Reviewer: ```suggestion
The sample code imports the [Pre-Generated Private Key](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) to the [`xKeychain`](https://github.com/ava-labs/avalanchejs/blob/46ce89f395133702320a77cba4bb9cb818b48fe8/examples/avm/buildExportTx-cchain-avax.ts#L31) to obtain signers. This means `X-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p` is the `balance` holder and will pay the `fee` to export AVAX.
```

### Step 5 - Run the Export and Import scripts

Between X/P/C chains there is shared memory. First, tokens are exported from the source chain to the shared memory, then imported by the destination chain.

To complete a transfer from the X-Chain to the C-Chain , we must run both the Export and Import scripts.

By default, the scripts use the AvalancheJS constant, [`DefaultLocalGenesisPrivateKey`](https://github.com/ava-labs/avalanchejs/blob/master/examples/avm/buildExportTx-cchain-avax.ts#L30) as `privKey` to sign transactions.

```js
const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
```

The bech32 address which derived from this private key is : `X-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p`

That key is pre-funded on the local network when using Avalanche Network Runner and has the following AVAX in the genesis block/vertex of each chain on the primary subnet:

- X-Chain: 300000000000000000 nAVAX

- P-Chain: 30000000000000000 nAVAX

- C-Chain: 50000000000000000000000000 nAVAX

```zsh
avalanchejs $ ts-node examples/avm/buildExportTx-cchain-avax.ts
Success! TXID: Rgg2412kaczRYC3taasvG6bYoqG7tBQG6WfacNdumKDKsVWpF

> avalanchejs $ ts-node examples/evm/buildImportTx-xchain.ts

Reviewer: buildImportTx-xchain.ts was never introduced until now. Please mention that above changes should be applied to both buildExportTx-cchain-avax.ts and buildImportTx-xchain.ts before running them.

Success! TXID: r2yYqcnCJcdeV5gddZ8NUoG5ZD3Ef7DxbkiE9xn4RxFcDdMd1
```

![Image for post](/img/ajs-getTx.png)

> That's it! You've transferred AVAX from the X-Chain to C-Chain! Now you can use them to deploy smart contracts on C-Chain.

Reviewer: ```suggestion
That's it! You've transferred AVAX from the X-Chain to C-Chain! 
```

You can retrieve the transaction data by running the following:

```zsh

> avalanchejs $ ts-node examples/avm/getTx.ts

Author: negative

Every example script will require some kind of modification

# Transfers AVAX Tokens Between Chains
# Cross Chain Transfers

## Introduction

This article shows how to transfer AVAX tokens programmatically between any two chains of the Primary Network.

> If you are looking for how to transfter AVAX tokens using the web wallet, please check out [this article](https://support.avax.network/en/articles/6169872-how-to-make-a-cross-chain-transfer-in-the-avalanche-wallet).

Reviewer: @Julian-dev28 if looks that you overwrote my changes. Please revert them.

Author: Resolved
https://github.com/ava-labs/avalanche-docs/pull/782/commits/b4ca846c0b718df3839a23ee05362b7e176b8ab1

  unsignedTx: {
    networkID: 5,
    blockchainID: '2JVSBoinj9C2J33VntvzYtVJNZdN2NKiwwKjcumHUWEb5DbBrm',
    outputs: [ [Object] ],
    inputs: [ [Object], [Object] ],
    memo: '0x41564d207574696c697479206d6574686f64206275696c644578706f7274547820746f206578706f7274204156415820746f2074686520432d436861696e2066726f6d2074686520582d436861696e',
    destinationChain: 'yH8D7ThNJkxmtkuv2jgBa4P1Rn3Qpr4pPr7QYNfcdoS6k6HWp',
    exportedOutputs: [ [Object] ]
  },
  credentials: [
    {
      fxID: 'spdxUxVJQbX85MGxMHbKw1sHxMnSqJ3QBzDyDYEP3h6TLuxqQ',
      credential: [Object]

>     },

Reviewer: are there two elements with the same value under `credentials`?

Author: confirmed, yes there is
